src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,186 @@
+            tail.next = newNode;
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index out of bound: " + index);
+        }
+        Node<T> newNode = new Node<>(null, value, null);
+        if (index == 0) {
+            if (this.isEmpty()) {
+                tail = head = newNode;
+            } else {
+                head.prev = newNode;
+                newNode.next = head;
+                head = newNode;
+            }
+        } else if (index == size) {
+            newNode.prev = tail;
+            newNode.prev.next = newNode;
+            tail = newNode;
+        } else {
+            Node<T> currentNode = head;
+            for (int i = 0; i < index - 1; i++) {
+                currentNode = currentNode.next;
+            }
+            newNode.next = currentNode.next;
+            if (currentNode.next != null) {
+                currentNode.next.prev = newNode;
+            }
+            currentNode.next = newNode;
+            newNode.prev = currentNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T listElement : list) {
+            add(listElement);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index out of bound: " + index);
+        }
+        if (this.isEmpty()) {
+            return null;
+        }
+
+        if (index == 0) {
+            return this.head.value;
+        } else if (index == size - 1) {
+            return this.tail.value;
+        } else {
+            Node<T> currentNode = this.head;
+            for (int i = 0; i < index; i++) {
+                currentNode = currentNode.next;
+            }
+            return currentNode.value;
+        }
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index out of bound: " + index);
+        }
+        if (this.isEmpty()) {
+            return null;
+        }
+        Node<T> currentNode = this.head;
+        for (int i = 0; i < index; i++) {
+            currentNode = currentNode.next;
+        }
+        T oldValue = currentNode.value;
+        currentNode.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index out of bound: " + index);
+        }
+        if (this.isEmpty()) {
+            return null;
+        }
+        Node<T> currentNode = this.head;
+        if (index == 0) {
+            this.head = currentNode.next;
+            if (this.head != null) {
+                this.head.prev = null;
+            }
```

Comment:
When removing the first element, ensure that the `tail` is also updated if the list becomes empty after removal. This will maintain the integrity of the list when it transitions from having one element to being empty.